### ISSUE TITLE
Wayland support

### ISSUE THREAD

**Issue by DavidLapous** [libvibrant/vibrantLinux#27]
I don't know if vibrantLinux is supposed to work on wayland, but it is not working on my computer :/.
Is wayland support possible or planned ?
Thanks for your work.

**Comment by zee-mzha** [comment]
Unfortunately I don't think this is possible. Nvidia has no version of their api that supports Wayland and the other method used for AMD relies on X.

**Comment by Scrumplex** [comment]
After some research about the AMD part I think it's currently not possible, as there is no standard API for our use-case.

If anyone wants to continue research, first of all the code in question belongs to [libvibrant](https://gitlab.com/Scrumplex/vibrant). The X11 implementation of the AMD part utilizes (lib-) XRandR and can be found [here](https://gitlab.com/Scrumplex/vibrant/-/blob/master/vibrant/src/ctm.c). 

As XRandR just exposes the Linux DRM properties `CTM` (from the CRT Controller/CTRC) to us, we just need a way to access those properties on various Wayland implementations. The closest *"standard"* API I could find was [this implementation in (GNOME) Mutter](https://wiki.gnome.org/Initiatives/Wayland/Gaps/DisplayConfig). I couldn't find anything in Weston nor wlroots. I also could not find anything in KWin (though, that might be because I don't really know it's structure).

If anyone continues / wants to continue research, I would be happy to talk with you. You can find [contact information on my website](https://scrumplex.net/)

**Comment by miermontoto** [comment]
I'm sorry to necro this, but this is one of the first results that google returns when searching for "vibrance linux".

In my case, if you use vibrantLinux on X11 and then log out and boot up Wayland, the saturation levels are somewhat mantained, so there must be a way to directly change the saturation on Wayland.

**Comment by Scrumplex** [comment]
I was playing around with this a few months ago and it might be possible, yes. Though it will probably require root privileges. I can look at this again. Maybe I can get this running

**Comment by miermontoto** [comment]
Thanks. I lack the technical knowledge to but if I can be of any help hit me up.

**Comment by Scrumplex** [comment]
I just tried doing this, and it seems hard to do yet again. I am not sure if there will be a way to do this in the future, without compositor-specific support. Not sure if that's viable, but it might be possible to use something like gamescope to achieve something like this. I don't know much about gamescope's codebase though. But maybe someone else can take this idea as an opportunity to work on this

**Comment by bunkbail** [comment]
> In my case, if you use vibrantLinux on X11 and then log out and boot up Wayland, the saturation levels are somewhat mantained, so there must be a way to directly change the saturation on Wayland.
This doesn't work with AMD cards. I was hoping it to work since CSGO run way smoother on Wayland than on X11.

**Comment by bunkbail** [comment]
Just leaving this as an idea: if your game can use Vulkan, you can use vkbasalt alongside a shader that does the same thing as vibrantLinux. CSGO can be used through zink to run it as Vulkan, but the performance still leaves much to be desired.

**Comment by miermontoto** [comment]
> This doesn't work with AMD cards. 

I'm running a 6900XT.

> if your game can use Vulkan

my problem is that my screen isn't properly tuned, so colors look really muted, it isn't game-specific.

**Comment by Scrumplex** [comment]
Since then I have implemented this for the Steam Deck using [vibrantDeck](https://github.com/libvibrant/vibrantDeck). The only technical requirement for it to work is actually only using Gamescope with the DRM backend (i.e. it only works if you run Gamescope as your primary compositor, not as a nested compositor).

No other Wayland compositor has a similar interface to Gamescope's so even a year later, this is still not possible to do.

I personally use Sway so if there is enough interest I can try to implement some interface for Sway. Though I can't guarantee that it will be merged upstream.

**Comment by Scrumplex** [comment]
vibrantDeck only works if gamescope is running as the primary compositor. It sadly will not work when it's running in nested mode

**Comment by Scrumplex** [comment]
I think it might be possible to use reshade to achieve what vibrantLinux does on a per-game basis. Gamescope also has a reshade interface now, so you might be able to get something working with it. But I don't know anything about writing shaders :(

**Comment by Scrumplex** [comment]
> So for vibrantDeck to work on desktop Linux basically it wouldn't be possible since the compositor is either X11 or Wayland, right?

vibrantDeck won't work on X11, yes. Wayland is just a protocol, so it depends on the implementation here. Gamescope is a Wayland compositor that exposes some color management options that we make use of in vibrantDeck. I don't know about any other compositor that exposes these options.

> Also, maybe there is a new way to implement vibrantDeck for the nested mode? Or maybe through XWayland?

Not really. Gamescope uses properties in the display driver using DRM. A nested window can not do that, as it's not the DRM master.

In other words, Gamescope can only set these values for a display output, not for a region on the screen.

> Either way, it would be a vibrance effect, and not really the digital vibrance.

It's the same thing ^^

All the GPU driver does is transform the RGB values. Your display won't show you now colors with vibrantLinux/vibrantDeck. It's really just a color transformation and a (reshade) shader should be able to do the same thing, but on a graphics level instead of a driver level

**Comment by Scrumplex** [comment]
Wayland color-management protocol is primarily intended for managing different color spaces on Wayland. You *might* be able to use this protocol to apply a custom color space using Gamescope, but I don't think this use-case is supported yet. Note that the protocol that was merged in Weston isn't merged into mainline wayland-protocols yet.

**Comment by Tiagoquix** [comment]
Great news, gamers. Great news.

https://www.phoronix.com/news/Wayland-Color-Management-Nears

**Comment by prateekshukla1108** [comment]
> Great news, gamers. Great news.
> 
> https://www.phoronix.com/news/Wayland-Color-Management-Nears

Finally !

**Comment by nahla-nee** [comment]
I might be missing something here, but I don't think that this protocol would do anything for this project.

per the xml spec at: https://gitlab.freedesktop.org/swick/wayland-protocols/-/blob/color/staging/color-management/color-management-v1.xml?ref_type=heads

> The aim of the color management extension is to allow clients to know the color properties of outputs, and to tell the compositor about the properties of their content on surfaces.

This spec doesn't have (as far as I can tell) anything in regards to managing these properties on behalf of another client, let alone the whole system, nor should it really if it were to conform to to Wayland's security standards. I think if Wayland support is achieved it would require maintaining a fork of the popular compositors/compositor bases (kwin, mutter, and wlroots), which I don't think is very realistic.

**Comment by Tremeschin** [comment]
Potential path for nvidia cards _(didn't find mentions anywhere, sorry if known)_:

As far as I understood, in the [`nvidia-open`](https://github.com/NVIDIA/open-gpu-kernel-modules) repository there's an `ioctl()` [SetDpyAttribute](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/25bef4626e6c5ccf5b433e1c22b6b1bd59e6f1bd/src/nvidia-modeset/interface/nvkms-api.h#L232) interface with a [Digital Vibrance](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/25bef4626e6c5ccf5b433e1c22b6b1bd59e6f1bd/src/nvidia-modeset/interface/nvkms-api.h#L2651) enum option through `/dev/nvidia-modeset`. This code seems to be synced with the proprietary driver, and parts of it are also found in `nvidia-dkms` at `/usr/src/nvidia-570.86.16/`, at least on Arch.

I'm not familiar with kernel modules to make further progress, but maybe it's possible to have a tiny code that interfaces with `nvidia.modeset` and sends a [NvKmsSetDpyAttributeParams](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/25bef4626e6c5ccf5b433e1c22b6b1bd59e6f1bd/src/nvidia-modeset/interface/nvkms-api.h#L2785) with the right command, attribute, and desired value? Would probably need building their [ioctl structs](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/25bef4626e6c5ccf5b433e1c22b6b1bd59e6f1bd/src/nvidia-modeset/interface/nvkms-ioctl.h#L47), initializing, querying all devices and displays etc. All-monitors should do as POC.

Otherwise, they seem to be aware and planning to eventually support it through NVML, as seen in the [nvidia-settings](https://github.com/NVIDIA/nvidia-settings/blob/6c755d9304bf4761f2b131f0687f0ebd1fcf7cd4/src/libXNVCtrlAttributes/NvCtrlAttributesNvml.c#L1235) code, the intended future way. Call your kernel savy friends, maybe we could have it today as the setting persists from xorg! 🙂

**Comment by Tremeschin** [comment]
Good news, green people! I managed to apply nvidia digital vibrance without depending on any video server or root permissions, directly through `/dev/nvidia-modeset` with `ioctl` calls, following my last comment! 🙂

Proof of concept and extra info: https://github.com/Tremeschin/nVibrant (WIP)

Libvibrant devs, feel absolutely free to base off and improve my work

**Comment by nahla-nee** [comment]
> Good news, green people! I managed to apply nvidia digital vibrance without depending on any video server or root permissions, directly through `/dev/nvidia-modeset` with `ioctl` calls, following my last comment! 🙂
> 
> Proof of concept and extra info: https://github.com/Tremeschin/nVibrant (WIP)
> 
> Libvibrant devs, feel absolutely free to base off and improve my work

That's awesome! We totally appreciate the effort. I can't speak for @Scrumplex, but when development of this project initially started I used to have an nvidia, and I no longer do so I can't really test it (neither does scrumplex afaik), but at a cursory glance it seems good!

I've personally moved on from this project and haven't maintained it in a long time. I don't know if scrumplex is still interested in it or has the time for it, but if you submit a pr that allows this mode to be used (set to opt-out by default on the gui for backwards compat) I will try to get it reviewed and merged in if he can't.

Thanks again for your work!

**Comment by Play-InTheWay** [comment]
For it to work as it does on x11 or Gamescope, we may need to propose a new wayland-protocol? Some months ago I write about a similar issue, so feel free to check it or complement it if necessary https://gitlab.freedesktop.org/wayland/wayland-protocols/-/issues/237

**Comment by augustoperes1** [comment]
Somehow I managed to get it working on Wayland (at least on Arch Linux with GNOME 49 and above). Basically, what I did was:

1. I installed Arch Linux with GNOME and GNOME Flashback (a session that mimics the old GNOME 2, and uses Metacity instead of Mutter);
2. In GDM, I selected GNOME Flashback;
3. I installed vibrantLinux using Flatpak;
4. I installed GNOME Tweaks;
5. In GNOME Tweaks, I added vibrantLinux to the applications that will start together with GNOME;
6. I configured vibrantLinux the way I use it;
7. I logged out of the GNOME Flashback session and then logged into the GNOME Shell 49 session with Wayland.

This worked for me, at least on Arch Linux with GNOME 49. I would appreciate it if you could test it if possible.

The only limitation is that you need to repeat this process every time you restart your computer.
